event-emitter/lib/emitter.js

```diff
@@ -30,38 +22,30 @@ class MyEventEmitter extends EventEmitter {
       // until either the next event is sent on the channel or a receive
       // timeout has occurred.
       if (this.isShutdown) {
-        return channel.shutdown();
+        channel.shutdown();
+        return;
       }
 
       // Poll for data
-      return (
-        poll()
-          .then(e => {
-            // Timeout on poll, no data to emit
-            if (!e) {
-              return undefined;
-            }
-
-            const { event, ...data } = e;
-
-            // Emit the event
-            this.emit(event, data);
-
-            return undefined;
-          })
-
-          // Emit errors
-          .catch(err => this.emit('error', err))
-
-          // Schedule the next iteration of the loop. This is performed with
-          // a `setImmediate` to extending the promise chain indefinitely
-          // and causing a memory leak.
-          .then(() => setImmediate(loop))
-      );
+      channel.poll((err, e) => {
+        if (err) this.emit('error', err);
+        else if (e) {
+          const { event, ...data } = e;
+
+          // Emit the event
+          this.emit(event, data);
+        }
+        // Otherwise, timeout on poll, no data to emit
+
+        // Schedule the next iteration of the loop. This is performed with
+        // a `setImmediate` to yield to the event loop, to let JS code run
+        // and avoid a stack overflow.
+        setImmediate(loop);
+      });
     };
 
-    // Start the polling loop
-    loop();
+    // Start the polling loop on next iteration of the JS event loop to prevent zalgo.
```

Comment:
Not everyone going through the examples will know what zalgo is. I think we should replace this with something more intuitive.

Reply:
Agreed + I think this change actually isn't necessary the way the example is currently written.